File: app/src/main/java/com/appchamp/wordchunks/ui/packs/PacksActivity.java
package com.appchamp.wordchunks.ui.packs;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;

import com.appchamp.wordchunks.R;
import com.appchamp.wordchunks.data.PacksRealmHelper;
import com.appchamp.wordchunks.models.realm.Pack;
import com.appchamp.wordchunks.ui.levels.LevelsActivity;

import java.util.List;

import io.realm.Realm;
import me.everything.android.ui.overscroll.OverScrollDecoratorHelper;
import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper;

import static com.appchamp.wordchunks.util.Constants.EXTRA_PACK_ID;


public class PacksActivity extends AppCompatActivity {

    private Realm realm;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.act_packs_levels);

        TextView tvTitle = (TextView) findViewById(R.id.tvTitle);
        tvTitle.setText(getString(R.string.title_select_pack));
        RecyclerView rvPacks = (RecyclerView) findViewById(R.id.recyclerView);

        realm = Realm.getDefaultInstance();

        // This is the RecyclerView adapter which will display the list of packs
        List<Pack> packs = PacksRealmHelper.findAllPacks(realm);

        PacksAdapter adapter = new PacksAdapter(packs);
        rvPacks.setAdapter(adapter);
        rvPacks.setLayoutManager(new LinearLayoutManager(PacksActivity.this));
        rvPacks.setHasFixedSize(true);

        adapter.setOnItemClickListener((view, position) -> {
            Pack clickedPack = adapter.getItem(position);
            startLevelsActivity(clickedPack.getId());
        });

        OverScrollDecoratorHelper.setUpOverScroll(
                rvPacks, OverScrollDecoratorHelper.ORIENTATION_VERTICAL);
    }
    
    @Override
    protected void attachBaseContext(Context newBase) {
        super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase));
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        overridePendingTransition(R.anim.slide_from_left, R.anim.slide_to_right);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        realm.close(); // Remember to close Realm when done.
    }

    public void onBackArrowClick(View v) {
        super.onBackPressed();
        overridePendingTransition(R.anim.slide_from_left, R.anim.slide_to_right);
    }

    /**
     * Up navigation. Navigates from PacksActivity to LevelsActivity passing packId by the Intent.
     * @param packId pack id to be passed by the Intent.
     */
    private void startLevelsActivity(String packId) {
        Intent intent = new Intent(PacksActivity.this, LevelsActivity.class);
        intent.putExtra(EXTRA_PACK_ID, packId);
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(intent);
        finish();
        overridePendingTransition(R.anim.slide_from_right, R.anim.slide_to_left);
    }
}

File: app/src/main/java/com/appchamp/wordchunks/ui/packs/PacksActivity.kt
/*
 * Copyright 2017 Julia Kozhukhovskaya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.appchamp.wordchunks.ui.packs

import android.arch.lifecycle.Observer
import android.content.Context
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import com.appchamp.wordchunks.R
import com.appchamp.wordchunks.realmdb.models.realm.Pack
import com.appchamp.wordchunks.ui.BaseActivity
import com.appchamp.wordchunks.ui.levels.LevelsActivity
import com.appchamp.wordchunks.util.Constants.EXTRA_PACK_ID
import com.franmontiel.localechanger.LocaleChanger
import io.realm.RealmResults
import kotlinx.android.synthetic.main.act_packs_levels.*
import kotlinx.android.synthetic.main.titlebar.*
import me.everything.android.ui.overscroll.OverScrollDecoratorHelper
import org.jetbrains.anko.clearTop
import org.jetbrains.anko.intentFor


class PacksActivity : BaseActivity<PacksViewModel>() {

    override val viewModelClass = PacksViewModel::class.java

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.act_packs_levels)
        llPacksLevels.setBackgroundResource(R.drawable.gradient_packs)
        tvTitle.text = getString(R.string.title_select_pack)
        rvList.layoutManager = LinearLayoutManager(this)
        rvList.setHasFixedSize(true)

        // Sets smooth scroll effect.
        OverScrollDecoratorHelper.setUpOverScroll(
                rvList, OverScrollDecoratorHelper.ORIENTATION_VERTICAL)

        imgBackArrow.setOnClickListener { onBackPressed() }

        subscribeToModel()
    }

    override fun attachBaseContext(newBase: Context) {
        super.attachBaseContext(LocaleChanger.configureBaseContext(newBase))
    }

    private fun subscribeToModel() {
        // Observe updates to our LiveData packs.
        viewModel
                .getLivePacks()
                .observe(this, Observer<RealmResults<Pack>> {
                    // update UI
                    val adapter = PacksLevelsAdapter<Pack> {
                        // Navigates to the LevelsActivity passing pack id in the Intent.
                        startLevelsActivity(it.id)
                    }

                    it?.toList()?.let { packs -> adapter.updateItems(packs) }
                    rvList.adapter = adapter
                })

        // Scroll RecyclerView to the last "current", or "solved" level item.
        // indexOfLast gets last index, or -1 if the list does not contain that item.
        rvList.smoothScrollToPosition(viewModel.getLastPackPos())
    }

    /**
     * Navigates back to MainActivity.
     */
    override fun onBackPressed() {
        super.onBackPressed()
        overridePendingTransition(R.anim.slide_from_left, R.anim.slide_to_right)
    }

    /**
     * Navigates back to MainActivity.
     */
    private fun startLevelsActivity(packId: String) {
        startActivity(intentFor<LevelsActivity>(EXTRA_PACK_ID to packId).clearTop())
        finish()
        overridePendingTransition(R.anim.slide_from_right, R.anim.slide_to_left)
    }
}